### ISSUE TITLE
Lesson on automating with CICD or Jenkins

### ISSUE THREAD

**Issue by dgjnpr** [nre-learning/antidote#153]
Hi,

I propose a CICD 3 stage lesson using Gitlab. This could be done with Git{Hub,Lab} + Jenkins but that requires additional plumbing and explaining (i.e. how the plumbing works).

1. Explain the setup - git repo of Ansible that applies a config, gitlab-ci conf
2. Create a branch, make config changes, use "--dryrun" in MR to test
3. How to review MR, approve and clear source branch, optionally update tag/version number

This is a more practical "how do I do CICD for networking". Prereqs are Ansible and it's "--dryrun" feature, basic git.

-- 
Cheeers
Dg

**Comment by Mierdin** [comment]
See #151 - you guys may want to collaborate. Could maybe use two lessons, but one of you focus on Gitlab, and the other dive deeper into Jenkins? Up to you.

FYI, I started looking into expanding the existing Git lesson with a Gitlab container. You can still have a separate lesson (which I think it's a good idea since you're going well beyond Git basics) but just FYI since everything in Antidote needs to be self-contained (i.e. no Github)

**Comment by Mierdin** [comment]
> Not sure how we avoid conflicts @Mierdin?

Can you clarify? Conflicts with what? In case you're referring to the lesson ID, don't worry about it. I'll make sure it's unique before it's merged in the unlikely event it isn't already.

By the way, even though you all are working on similar lesson ideas, it doesn't mean you all have to build the same lesson. Just figure it would be worth coordinating so you can tackle topics the way you want without stepping on each others toes. That might mean you all build different labs in the same lesson (if your content focuses on the same general concepts), or it might not. That's up to you to plan out

> Since Gitlab is an absolute resource hog, I would propose a simpler (and smaller) SCM, like GOGS, which offers a Github-like interface, and is much lighter than Gitlab.

Keep in mind we're running network devices :smile: so resource demands are fairly high already. I definitely appreciate keeping it light where it makes sense, as long as it doesn't detract from the learning experience by teaching something that's totally different from what the user might see in the real world. GOGS looks pretty slick and Github-like as you mention, so just make sure in the content/guides you do a good job of mapping GOGS concepts to Github concepts, because let's face it, that's where this knowledge is likely to get applied.

**Comment by ntwrkguru** [comment]
I agree @Mierdin. We've used GOGS for a customer solution before, so I'm confident that it's sufficient and won't detract.

**Comment by Mierdin** [comment]
@ntwrkguru By the way, I created https://github.com/nre-learning/syringe/pull/53 to add lesson ID uniqueness checks to the syringe command-line tool. Shortly after the next release I'll follow up with https://github.com/nre-learning/antidote/issues/158 and get it added to CI. That way you know in your PRs that your lesson ID is unique.

**Comment by Mierdin** [comment]
Hey @ntwrkguru and @dgjnpr, any progress on this one? Anything I can do to help?

**Comment by ntwrkguru** [comment]
Swamped with customer projects, but should get back to this next week. Sorry! Day job comes first!

Sent from my iPhone

On Feb 10, 2019, at 12:35 AM, Matt Oswalt <notifications@github.com<mailto:notifications@github.com>> wrote:


Hey @ntwrkguru<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ntwrkguru&d=DwMCaQ&c=HAkYuh63rsuhr6Scbfh0UjBXeMK-ndb3voDTXcWzoCI&r=y8f5-ZKbifhgVsSVihKTuG2bbsBaiqELScS4YCpX8v8&m=RdVlwLhGPNAB2T0Mg2v0EMKaBabKC3nltgZMaPFxtXQ&s=aLxtpLJQsy4s_jjYsLDmPWyJdS2_Hf3kv-2QF97wwok&e=> and @dgjnpr<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dgjnpr&d=DwMCaQ&c=HAkYuh63rsuhr6Scbfh0UjBXeMK-ndb3voDTXcWzoCI&r=y8f5-ZKbifhgVsSVihKTuG2bbsBaiqELScS4YCpX8v8&m=RdVlwLhGPNAB2T0Mg2v0EMKaBabKC3nltgZMaPFxtXQ&s=lSR2RSRmwCL0KAw0YfAtxu7eJgo4aKRN2cjA_s-Ynt0&e=>, any progress on this one? Anything I can do to help?
